src/polymer/polymer2-element-scanner.ts

```diff
@@ -405,10 +417,65 @@ class ElementVisitor implements Visitor {
       const element = this._possibleElements.get(className);
       if (element) {
         element.className = className;
-        element.tagName = tagName;
+        element.tagName = this.getTagNameFromExpression(tagName);
         this._elements.add(element);
       }
     }
     return Array.from(this._elements);
   }
+
+  getTagNameFromExpression(expression: TagNameExpression): string|undefined {
+    if (expression.type === 'string-literal') {
+      return expression.value;
+    }
+    const element = this._possibleElements.get(expression.className) ||
+        Array.from(this._elements)
+            .find((e) => e.className === expression.className);
+    if (!element) {
+      this._warnings.push({
+        code: 'cant-determine-element-tagname',
+        message:
+            `Couldn't dereference the class name ${expression.className} here.`,
+        severity: Severity.WARNING,
+        sourceRange: expression.classNameSourceRange
+      });
+      return;
+    }
+    return element.tagName;
+  }
+
+  getTagNameExpression(expression: estree.Node): TagNameExpression|undefined {
+    const tryForLiteralString = astValue.expressionToValue(expression);
+    if (tryForLiteralString != null &&
+        typeof tryForLiteralString === 'string') {
+      return {
+        type: 'string-literal',
+        value: tryForLiteralString,
+        sourceRange: this._document.sourceRangeForNode(expression)!
+      };
+    }
+    if (expression.type === 'MemberExpression') {
+      // Might be something like MyElement.is
+      const isPropertyNameIs = (expression.property.type === 'Identifier' &&
+                                expression.property.name === 'is') ||
+          (astValue.expressionToValue(expression.property) === 'is');
+      const className = astValue.getIdentifierName(expression.object);
+      if (isPropertyNameIs && className) {
+        return {
+          type: 'is',
+          className,
+          classNameSourceRange:
+              this._document.sourceRangeForNode(expression.object)!
+        };
+      }
+    }
+    this._warnings.push({
```

Comment:
+1, thanks for pushing for more functional and less side-effecty.

Introduced a new local type `OrWarning<T>` for this and getTagNameFromExpression